### PR TITLE
[core] merging formats by codecs, determining extension before download

### LINKED ISSUE #4069: --get-filename yields inaccurate extension when the format will be merged
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.05.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

running --get-filename with a format specifier that requires combining streams produces inaccurate results

i suspect this issue has been around for a while but youtube shorts force it to occur more often than most due to their dimensions

### Verbose log

```shell
C:\Users\ej> yt-dlp -vU -f bv[height<=720]+ba --force-ipv4 --sleep-interval 5 --max-sleep-interval 30 --no-continue --no-overwrites --no-check-certificate --no-mtime --no-warnings https://youtu.be/nfsCkGJUl_8 --get-filename -o "%(title)s-%(id)s.%(ext)s"
[debug] Command-line config: ['-vU', '-f', 'bv[height<=720]+ba', '--force-ipv4', '--sleep-interval', '5', '--max-sleep-interval', '30', '--no-continue', '--no-overwrites', '--no-check-certificate', '--no-mtime', '--no-warnings', 'https://youtu.be/nfsCkGJUl_8', '--get-filename', '-o', '%(title)s-%(id)s.%(ext)s']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d523] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 5.0-essentials_build-www.gyan.dev (setts), ffprobe 5.0-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [youtube] Extracting URL: https://youtu.be/nfsCkGJUl_8
[youtube] nfsCkGJUl_8: Downloading webpage
[youtube] nfsCkGJUl_8: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[info] nfsCkGJUl_8: Downloading 1 format(s): 136+251
Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.mp4
C:\Users\ej> yt-dlp -vU -f bv[height<=720]+ba --force-ipv4 --sleep-interval 5 --max-sleep-interval 30 --no-continue --no-overwrites --no-check-certificate --no-mtime --no-warnings https://youtu.be/nfsCkGJUl_8 -o "%(title)s-%(id)s.%(ext)s"  [debug] Command-line config: ['-vU', '-f', 'bv[height<=720]+ba', '--force-ipv4', '--sleep-interval', '5', '--max-sleep-interval', '30', '--no-continue', '--no-overwrites', '--no-check-certificate', '--no-mtime', '--no-warnings', 'https://youtu.be/nfsCkGJUl_8', '-o', '%(title)s-%(id)s.%(ext)s']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d523] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 5.0-essentials_build-www.gyan.dev (setts), ffprobe 5.0-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[debug] [youtube] Extracting URL: https://youtu.be/nfsCkGJUl_8
[youtube] nfsCkGJUl_8: Downloading webpage
[youtube] nfsCkGJUl_8: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[info] nfsCkGJUl_8: Downloading 1 format(s): 136+251
[debug] Invoking http downloader on "https://rr2---sn-3n4pcxg-pjul.googlevideo.com/videoplayback?expire=1655240708&ei=pKOoYr3LDKyHir4Pkc2K6A8&ip=199.87.203.248&id=o-AEyy8G9d_Wpeo2v4iabrZbBLRbQ5oLYGIfFR_w05LeCG&itag=136&source=youtube&requiressl=yes&mh=zx&mm=31%2C29&mn=sn-3n4pcxg-pjul%2Csn-vgqsrnsd&ms=au%2Crdu&mv=m&mvi=2&pl=21&ctier=SH&initcwndbps=967500&vprv=1&mime=video%2Fmp4&gir=yes&clen=1708677&dur=41.291&lmt=1655116620713722&mt=1655218620&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5432434&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cctier%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgCVVjHe0-WOU_2ioZ3UPokwAqJ5vDBvHxqCW29QOsoYUCIQDmj5AMM8ZTL9SLl7Nf9coghK9E2kIw5PDJH1vAEN-RyQ%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALl0e4tm-ohMXN_sSrvSVsGbM9j_mSCCN2QD8pb0ctvDAiBNzE8E2yruR3J7rYygG22EeoZRzjEEHlS6JgLwWIscbA%3D%3D"
[download] Sleeping 9.97 seconds ...
[debug] File locking is not supported on this platform. Proceeding without locking
[download] Destination: Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f136.mp4
[download] 100% of 1.63MiB in 00:00
[debug] Invoking http downloader on "https://rr2---sn-3n4pcxg-pjul.googlevideo.com/videoplayback?expire=1655240708&ei=pKOoYr3LDKyHir4Pkc2K6A8&ip=199.87.203.248&id=o-AEyy8G9d_Wpeo2v4iabrZbBLRbQ5oLYGIfFR_w05LeCG&itag=251&source=youtube&requiressl=yes&mh=zx&mm=31%2C29&mn=sn-3n4pcxg-pjul%2Csn-vgqsrnsd&ms=au%2Crdu&mv=m&mvi=2&pl=21&ctier=SH&initcwndbps=967500&vprv=1&mime=audio%2Fwebm&gir=yes&clen=539480&dur=41.321&lmt=1655116737253115&mt=1655218620&fvip=1&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5432434&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cctier%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgPyukwntpjMQRkbkI2yP4cJbyA8E2VXSAvj3w5cHZF2kCIGl1Dw0gkQV_zYZ1MtTrb7Z2-HjpZmIA_3KAR6Vt3Npq&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhALl0e4tm-ohMXN_sSrvSVsGbM9j_mSCCN2QD8pb0ctvDAiBNzE8E2yruR3J7rYygG22EeoZRzjEEHlS6JgLwWIscbA%3D%3D"
[download] Sleeping 9.81 seconds ...
[download] Destination: Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f251.webm
[download] 100% of 526.84KiB in 00:00
[Merger] Merging formats into "Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.mkv"
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f136.mp4" -i "file:Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f251.webm" -c copy -map "0:v:0" -map "1:a:0" -movflags "+faststart" "file:Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.temp.mkv"
Deleting original file Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f136.mp4 (pass -k to keep)
Deleting original file Why You Should Always Wash Your Vegetables-nfsCkGJUl_8.f251.webm (pass -k to keep)
C:\Users\ej>
```

### PR DESCRIPTION
**Pull request by selfisekai**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

- determine the file extension of a merged video *before* downloading
- better file extension matching

fixes #4069

### CODE REVIEW

**Comment by megapro17** [comment]
mp4 can support vp9

**Comment by nao20010128nao** [comment]
didn't know VP9+Opus is possible for MP4
```
$ yt-dlp -f 'bv[vcodec=vp9]+ba[acodec=opus]' --remux mp4 https://www.youtube.com/watch?v=cmppl9gf2Qw
[youtube] cmppl9gf2Qw: Downloading webpage
[youtube] cmppl9gf2Qw: Downloading android player API JSON
[info] cmppl9gf2Qw: Downloading 1 format(s): 247+251
[download] Destination: 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].f247.webm
[download] 100% of 69.97MiB in 00:00                  
[download] Destination: 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].f251.webm
[download] 100% of 7.79MiB in 00:00                  
[Merger] Merging formats into "【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].webm"
Deleting original file 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].f251.webm (pass -k to keep)
Deleting original file 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].f247.webm (pass -k to keep)
[VideoRemuxer] Remuxing video from webm to mp4; Destination: 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].mp4
Deleting original file 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].webm (pass -k to keep)
$ ffprobe 【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】\ \[cmppl9gf2Qw\].mp4 
ffprobe version 4.3.2-0+deb11u1ubuntu1 Copyright (c) 2007-2021 the FFmpeg developers
  built with gcc 10 (Ubuntu 10.2.1-20ubuntu1)
(omit)
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '【マインクラフト】危険すぎる巨大サイコロを開封する【ゆっくり実況】 [cmppl9gf2Qw].mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf58.45.100
  Duration: 00:08:13.02, start: 0.000000, bitrate: 1326 kb/s
    Stream #0:0(eng): Video: vp9 (Profile 0) (vp09 / 0x39307076), yuv420p(tv, bt709), 1280x720, 1188 kb/s, SAR 1:1 DAR 16:9, 30 fps, 30 tbr, 16k tbn, 16k tbc (default)
    Metadata:
      handler_name    : VideoHandler
    Stream #0:1(eng): Audio: opus (Opus / 0x7375704F), 48000 Hz, stereo, fltp, 129 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
```

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -6616,3 +6616,41 @@ def join_nonempty(*values, delim='-', from_dict=None):
     if from_dict is not None:
         values = map(from_dict.get, values)
     return delim.join(map(str, filter(None, values)))
+
+def get_compatible_ext(vcodecs, acodecs, vexts, aexts):
+    assert len(vcodecs) == len(vexts) and len(acodecs) == len(aexts)
+
+    if max(len(acodecs), len(vcodecs)) > 1:
+        # TODO: any other format allows this?
+        output_ext = 'mkv'
```
Comment (pukkandan) [inline yt_dlp/utils.py:6625]:
```suggestion
        return 'mkv'
```

**Comment by gamer191** [comment]
Please add `Fixes https://github.com/yt-dlp/yt-dlp/issues/4069` to the PR description

File: yt_dlp/YoutubeDL.py [hunk 2]
```diff
@@ -3067,41 +3065,16 @@ def existing_video_file(*filepaths):
...
-                                'Requested formats are incompatible for merge and will be merged into mkv')
                         if (info_dict['ext'] == 'webm'
                                 and info_dict.get('thumbnails')
                                 # check with type instead of pp_key, __name__, or isinstance
                                 # since we dont want any custom PPs to trigger this
                                 and any(type(pp) == EmbedThumbnailPP for pp in self._pps['post_process'])):  # noqa: E721
                             info_dict['ext'] = 'mkv'
-                            self.report_warning(
-                                'webm doesn\'t support embedding a thumbnail, mkv will be used')
+                            self.report_warning('webm doesn\'t support embedding a thumbnail, mkv will be used')
```
Comment (pukkandan) [inline yt_dlp/YoutubeDL.py:3077]:
```suggestion
                            self.report_warning(
                                'webm doesn\'t support embedding a thumbnail, mkv will be used')
```

**Pull request merged**